### PR TITLE
feat: add server-side pagination sort and search for centers component

### PR DESCRIPTION
**Pull request by punwai**
## Description
- Add Server-side pagination, sort and search for centers component
- Refactored code to meet current guidelines

## Related issues and discussion
#234

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by punwai** [comment]
If you review the groups component and there are issues, please don't review this one, as it will also have similar issues :)

**Comment by punwai** [comment]
The search for the name and externalId, which is specified on the API, does not work. This is because the API endpoint for name and externalID queries do not work. Please tell me if I am wrong in any way, but here are two queries which should return at least 1 item, but returns nothing.

https://demo.openmf.org/fineract-provider/api/v1/centers?tenantIdentifier=default&pretty=true&paged=true&name=Santa%20Cruz

https://demo.openmf.org/fineract-provider/api/v1/centers?tenantIdentifier=default&pretty=true&paged=true&externalId=Bqmto

**Comment by punwai** [comment]
https://demo.openmf.org/api-docs/apiLive.htm#centers_list

File: src/app/centers/centers.service.ts [hunk 1]
```diff
@@ -1,17 +1,45 @@
...
+    let httpParams = new HttpParams()
+      .set('offset', offset.toString())
+      .set('limit', limit.toString())
+      .set('sortOrder', sortOrder)
+      .set('orderBy', orderBy)
+      .set('paged', 'true');
+    // filterBy: name, externalId
+    filterBy.forEach(function (filter: any) {
+      if (filter.value) {
+        console.log(filter.value);
```
Comment (abhaychawla) [inline src/app/centers/centers.service.ts]: Remove `console.log` statements.
```suggestion
```

File: src/app/centers/centers.component.html [hunk 2]
```diff
@@ -1,53 +1,60 @@
...
+      </td>
+    </ng-container>
+
+    <ng-container matColumnDef="officeName">
+      <th mat-header-cell *matHeaderCellDef mat-sort-header> Office Name </th>
+      <td mat-cell *matCellDef="let center"> {{ center.officeName }} </td>
+    </ng-container>
+
+    <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
+    <tr mat-row *matRowDef="let row; columns: displayedColumns;" [routerLink]="[ ]"></tr>
```
Comment (abhaychawla) [inline src/app/centers/centers.component.html]:
```suggestion
    <tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
```
Remove this as its not required for the time being.

File: src/app/centers/centers.component.scss [hunk 3]
```diff
@@ -1,27 +1,10 @@
...
+    margin-left: auto;
   }
-  
-  .mat-form-field {
-    font-size: 14px;
-    width: 100%;
+  .search-button{
+    min-width: 30px;
+    margin: 4px;
+
```
Comment (abhaychawla) [inline src/app/centers/centers.component.scss]: Remove extra line.
```suggestion
```

File: src/app/centers/centers.component.ts [hunk 4]
```diff
@@ -1,45 +1,132 @@
...
-import { Observable } from 'rxjs';
+/** Angular Imports */
+import { Component, OnInit, ViewChild, AfterViewInit } from '@angular/core';
+import { MatPaginator, MatSort, MatCheckbox } from '@angular/material';
+import { FormControl } from '@angular/forms';
+import { Router, ActivatedRoute } from '@angular/router';
 
+/** rxjs Imports */
+import { merge } from 'rxjs';
+import { tap, startWith, map, distinctUntilChanged, debounceTime} from 'rxjs/operators';
```
Comment (abhaychawla) [inline src/app/centers/centers.component.ts]:
```suggestion
import { tap, startWith, map, distinctUntilChanged, debounceTime } from 'rxjs/operators';
```

**Comment by abhaychawla** [comment]
> The search for the name and externalId, which is specified on the API, does not work. This is because the API endpoint for name and externalID queries do not work. Please tell me if I am wrong in any way, but here are two queries which should return at least 1 item, but returns nothing.
> 
> https://demo.openmf.org/fineract-provider/api/v1/centers?tenantIdentifier=default&pretty=true&paged=true&name=Santa%20Cruz
> 
> https://demo.openmf.org/fineract-provider/api/v1/centers?tenantIdentifier=default&pretty=true&paged=true&externalId=Bqmto

Raise a Fineract ticket and a corresponding issue linked to it here for this.

**Comment by abhaychawla** [comment]
@punwai Are you able to update the PR?

**Comment by abhaychawla** [comment]
Contributor is inactive, but merging the PR as the required changes do not impact the functionality and can be done as part of refactoring later.

**Pull request merged**